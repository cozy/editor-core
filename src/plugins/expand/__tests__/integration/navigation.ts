import { BrowserTestCase } from '@atlaskit/webdriver-runner/runner';

import {
  fullpage,
  editable,
  getDocFromElement,
} from '../../../../__tests__/integration/_helpers';
import {
  goToEditorTestingExample,
  mountEditor,
} from '../../../../__tests__/__helpers/testing-example-helpers';
import { selectors } from '../../../../__tests__/__helpers/page-objects/_expand';

import emptyExpandAdf from './__fixtures__/empty-expand.json';
import doubleExpand from './__fixtures__/double-expand.json';

const collapseExpandThenFocusTitle = async (page: any) => {
  await page.click(selectors.expandToggle);
  await page.click(selectors.expandTitleInput);
};

BrowserTestCase(
  'navigation.ts: pressing Backspace should delete an expand when cursor is inside content',
  { skip: ['edge', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: emptyExpandAdf,
      allowExpand: true,
    });

    await page.click(selectors.expandContent);
    await page.keys('Backspace');

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

BrowserTestCase(
  'navigation.ts: pressing Backspace should delete an expand when cursor is inside title',
  { skip: ['edge', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: emptyExpandAdf,
      allowExpand: true,
    });

    await collapseExpandThenFocusTitle(page);
    await page.keys('Backspace');

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

BrowserTestCase(
  'navigation.ts: pressing Enter should collapse an expand when cursor is inside title',
  { skip: ['edge', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: emptyExpandAdf,
      allowExpand: true,
    });

    await collapseExpandThenFocusTitle(page);

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

BrowserTestCase(
  'navigation.ts: when cursor is after a collapsed expand, pressing Backspace should focus the title',
  { skip: ['edge', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: emptyExpandAdf,
      allowExpand: true,
    });

    await collapseExpandThenFocusTitle(page);
    await page.keys(['ArrowDown', 'Backspace']);
    await page.keys('title'); // type into the open, not targeted at a field

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);

BrowserTestCase(
  'navigation.ts: when cursor is after two collapsed expands, pressing Backspace should focus the title of the second one',
  { skip: ['edge', 'safari', 'firefox'] },
  async (client: any, testName: string) => {
    const page = await goToEditorTestingExample(client);

    await mountEditor(page, {
      appearance: fullpage.appearance,
      defaultValue: doubleExpand,
      allowExpand: true,
    });

    await page.click('.ak-editor-expand__title-input[value="Second title"]');
    await page.keys(['ArrowDown', 'Backspace']);
    await page.keys('hello'); // type into the open, not targeted at a field

    const doc = await page.$eval(editable, getDocFromElement);
    expect(doc).toMatchCustomDocSnapshot(testName);
  },
);
