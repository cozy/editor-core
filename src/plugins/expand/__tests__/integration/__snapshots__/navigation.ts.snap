// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navigation.ts: pressing Backspace should delete an expand when cursor is inside content 1`] = `
Object {
  "content": Array [
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: pressing Backspace should delete an expand when cursor is inside title 1`] = `
Object {
  "content": Array [
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: pressing Enter should collapse an expand when cursor is inside title 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "__expanded": false,
        "title": "",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: when cursor is after a collapsed expand, pressing Backspace should focus the title 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "__expanded": false,
        "title": "title",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`navigation.ts: when cursor is after two collapsed expands, pressing Backspace should focus the title of the second one 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "text": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean mi nisl, venenatis eget auctor vitae, venenatis quis lorem. Suspendisse maximus tortor vel dui tincidunt cursus. Vestibulum magna nibh, auctor non auctor id, finibus vitae orci. Nulla viverra ipsum et nunc fringilla ultricies. Pellentesque vitae felis molestie justo finibus accumsan. Suspendisse potenti. Nulla facilisi. Integer dignissim quis velit quis elementum. Sed sit amet varius ante. Duis vestibulum porta augue eu laoreet. Morbi id risus et augue sollicitudin aliquam. In et ligula dolor. Nam ac aliquet diam.",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "attrs": Object {
        "__expanded": false,
        "title": "First title",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "attrs": Object {
        "__expanded": false,
        "title": "Second titlehello",
      },
      "content": Array [
        Object {
          "type": "paragraph",
        },
      ],
      "type": "expand",
    },
    Object {
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;
